lib/killbill/helpers/active_merchant/utils.rb

```diff
@@ -46,100 +49,65 @@ def write(string)
           end
         end
 
-        # Relies on the fact that hashes enumerate their values in the order that the corresponding keys were inserted (Ruby 1.9+)
         class BoundedLRUCache
 
-          def initialize(proc, max_size=10000)
-            @proc     = proc
+          def initialize(proc, max_size = 10000)
             @max_size = max_size
-
-            if defined?(JRUBY_VERSION)
-              @is_jruby  = true
-              @semaphore = nil
-
-              lru_cache = Class.new(java.util.LinkedHashMap) do
-                def initialize(max_size)
-                  super(max_size, 1.0, true)
-                  @max_size = max_size
-                end
-
-                # Note: renaming it to remove_eldest_entry won't work
-                def removeEldestEntry(eldest)
-                  size > @max_size
-                end
-              end.new(@max_size)
-              @data     = java.util.Collections.synchronizedMap(lru_cache)
-            else
-              @is_jruby  = false
-              @semaphore = Mutex.new
-              # TODO Pre-allocate?
-              @data      = {}
+            @data = ThreadSafe::Cache.new do |hash, key|
+              # mapping key -> value is constant for our purposes
+              set_value = hash.fetch_or_store key, value = proc.call(key)
+              store_key(key) if value.equal?(set_value) # very same object
+              set_value
             end
+            @keys = []
```

Comment:
yep, it's just that we need to do more under a lock in every use-case than a simple array operation
so that would change the "atomicity" of a write (removing when full size) but should be fine ... right?